tstoreds/keybook.go

```diff
@@ -0,0 +1,193 @@
+package tstoreds
+
+import (
+	"fmt"
+
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	ic "github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/textileio/go-textile-core/thread"
+	tstore "github.com/textileio/go-textile-core/threadstore"
+	"github.com/whyrusleeping/base32"
+)
+
+type dsKeyBook struct {
+	ds ds.Datastore
+}
+
+// Public and private keys are stored under the following db key pattern:
+// /threads/keys/<b32 thread id no padding>/<b32 log id no padding>/(pub|priv|read|follow)
+var (
+	kbBase       = ds.NewKey("/thread/keys")
+	pubSuffix    = ds.NewKey("/pub")
+	privSuffix   = ds.NewKey("/priv")
+	readSuffix   = ds.NewKey("/read")
+	followSuffix = ds.NewKey("/follow")
+)
+
+var _ tstore.KeyBook = (*dsKeyBook)(nil)
+
+// NewKeyBook returns a new key book for storing public and private keys
+// of (thread.ID, peer.ID) pairs with durable guarantees by store.
+func NewKeyBook(store ds.Datastore) (tstore.KeyBook, error) {
+	return &dsKeyBook{ds: store}, nil
+}
+
+// PubKey returns the public key of (thread.ID, peer.ID). The implementation
+// assumes the key is in the store with the exception that peer.ID is an
+// Identity multihash. If the public key can't be resolved, nil is returned.
+func (kb *dsKeyBook) PubKey(t thread.ID, p peer.ID) (ic.PubKey, error) {
+	key := dsKey(t, p, kbBase).Child(pubSuffix)
+
+	var pk ic.PubKey
+	if v, err := kb.ds.Get(key); err == nil {
+		if pk, err = ic.UnmarshalPublicKey(v); err != nil {
+			return nil, fmt.Errorf("store backed public key %s can't be unmarshaled: %w", key, err)
+		}
+	} else if err == ds.ErrNotFound {
+		pk, err = p.ExtractPublicKey()
```

Comment:
This case is a little weird. I think I removed it from the mem store version. Extracting and adding _within_ a get call feels like a black box... perhaps useful in the normal peerstore, but probably not what we want with threads. Checking if pubkey is not nil is useful to see if we have the log.

Reply:
Yeah, it's a bit magical. Also, seems unexpected that a read might eventually affect the store (an indirect a `.Put` is done). Gonna take it out.